components/proxy/src/main/java/com/hotels/styx/admin/handlers/ServiceProviderHandler.java

```diff
@@ -0,0 +1,127 @@
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.common.collect.ImmutableList;
+import com.hotels.styx.api.Eventual;
+import com.hotels.styx.api.HttpInterceptor;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.WebServiceHandler;
+import com.hotels.styx.routing.config.StyxObjectDefinition;
+import com.hotels.styx.routing.db.StyxObjectStore;
+import com.hotels.styx.routing.handlers.ProviderObjectRecord;
+import org.slf4j.Logger;
+
+import java.util.List;
+
+import static com.fasterxml.jackson.core.JsonParser.Feature.AUTO_CLOSE_SOURCE;
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
+import static com.hotels.styx.admin.handlers.UrlPatternRouter.placeholders;
+import static com.hotels.styx.api.HttpResponse.response;
+import static com.hotels.styx.api.HttpResponseStatus.NOT_FOUND;
+import static com.hotels.styx.api.HttpResponseStatus.NO_CONTENT;
+import static com.hotels.styx.api.HttpResponseStatus.OK;
+import static com.hotels.styx.infrastructure.configuration.json.ObjectMappers.addStyxMixins;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.stream.Collectors.joining;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Admin interface access to service provider information, esp. Health Check monitors.
+ */
+public class ServiceProviderHandler implements WebServiceHandler {
+
+    private static final Logger LOGGER = getLogger(ServiceProviderHandler.class);
+
+    private static final ObjectMapper YAML_MAPPER = addStyxMixins(new ObjectMapper(new YAMLFactory()))
+            .addMixIn(StyxObjectDefinition.class, ServiceProviderDefMixin.class)
+            .configure(FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .configure(AUTO_CLOSE_SOURCE, true);
+
+    private final UrlPatternRouter urlRouter;
+
+    public ServiceProviderHandler(StyxObjectStore<ProviderObjectRecord> providerDatabase) {
+        urlRouter = new UrlPatternRouter.Builder()
+                .get("/admin/service/providers", (request, context) -> {
+                    String output = providerDatabase.entrySet()
+                            .stream()
+                            .map(entry -> serialise(entry.getKey(), entry.getValue()))
+                            .collect(joining("\n"));
+
+                    return Eventual.of(response(output.length() == 0 ? NO_CONTENT : OK)
+                            .body(output, UTF_8)
+                            .build());
+
+                })
+                .get("/admin/service/provider/:providerName", (request, context) -> {
+                    String name = placeholders(context).get("providerName");
+
+                    try {
+                        String object = providerDatabase.get(name)
+                                .map(record -> serialise(name, record))
+                                .orElseThrow(ResourceNotFoundException::new);
```

Comment:
I see we have several of these ResourceNotFoundException classes.   Should we return the response from the map() call instead of having to throw an exception and capture it?

Reply:
Yes, much nicer.